### PR TITLE
Merging master into add-dns-challenge

### PR DESCRIPTION
**Pull request by captncraig**
Also adding client.UseDNSChallenge in order to allow library users to access the dns functionality.

### CODE REVIEW

File: acme/client.go [hunk 1]
```diff
@@ -611,3 +577,54 @@ func parseLinks(links []string) map[string]string {
...
+		hdr, err = getJSON(uri, &challengeResponse)
+		if err != nil {
+			return err
+		}
+	}
+}
+
+// UseDNSChallenge configures the client to use the provided DNSProvider
+// as it's only challenge
+func (c *Client) UseDNSChallenge(d DNSProvider) {
```
Comment (captncraig) [inline acme/client.go]: this is a change I made outside of just a merge.

File: acme/dns_challenge.go [hunk 2]
```diff
@@ -44,6 +44,10 @@ func (s *dnsChallenge) Solve(chlng challenge, domain string) error {
 		return err
 	}
 
+	// sleep a little bit to let the record propigate.
```
Comment (captncraig) [inline acme/dns_challenge.go]: This is the only other change I made besides merging. Temporary, but it makes things work until querying is implemented.

File: acme/client.go (hunk 1 above)
Comment (mholt) [inline acme/client.go]: I'd rather just stick to the ExcludeChallenges method, listing every challenge except DNS. (There's only 4 challenges total.)

Comment (captncraig) [inline acme/client.go]: Exclude doesn't give the capability I need. I have to give it the dns provider which has a ton of internal configuration.

**Comment by xenolf** [comment]
DNS is now in master :)

**Pull request closed without merging**